dask/optimize.py

```diff
@@ -427,6 +428,179 @@ def fuse_getitem(dsk, func, place):
                            lambda a, b: tuple(b[:place]) + (a[2], ) + tuple(b[place + 1:]))
 
 
+def fuse_reductions(dsk, keys=None, ave_width=2, max_depth_new_edges=None,
+                    max_height=None, max_width=None):
+    """ WIP. Probably broken.  Fuse tasks that form reductions.
+
+    This trades parallelism opportunities for faster scheduling by making
+    tasks less granular.
+
+    There are many options (and opinions) for what the parameterization
+    of this function should be.  For the sake of experimentation, I have
+    included many parameters as described below.  My gut feeling right
+    now is that we only need one parameter, `ave_width`, and to choose
+    a reasonable value of `max_depth_new_edges` based on `ave_width`
+    (this controls against pathologies while allowing desirable behavior).
+
+    This optimization operation is more general than "single input, single
+    output".  It applies to all reductions, so it may be "multiple input,
+    single output".  I have attempted to make this well-behaved and
+    robust against pathologies.  Notably, by allowing multiple inputs
+    (what I refer to as `edges` in the code), there may be parallelism
+    opportunities w.r.t. the edges that are otherwise not captured by
+    analyzing the fusible reduction tasks alone.  I added a cheap-to-
+    compute heuristic that is conservative; i.e., it will never
+    underestimate the degree of edge parallelism.  This heuristic
+    increases the value compared against `min_width`--a measure of
+    parallizability--which is one of the reasons I prefer `min_width`.
+
+    I think it will be easy for this operation to supercede `fuse`.
+    I'm ignoring task renaming for now.
+
+    Parameters
+    ----------
+    dsk: dict
+        dask graph
+    keys: list or set
+        Keys that must remain in the dask graph
+    ave_width: float
+        Limit for `width = num_nodes / height`, a good measure of
+        parallelizability.
+    max_depth_new_edges: int
+        Don't fuse if new dependencies are added after this many levels
+    max_height: int
+        Don't fuse more than this many levels
+    max_width: int
+        Don't fuse if total width is greater than this
+
+    """
+    if keys is not None and not isinstance(keys, set):
+        if not isinstance(keys, list):
+            keys = [keys]
+        keys = set(flatten(keys))
+
+    # Assign reasonable, not too restrictive defaults
+    if max_depth_new_edges is None:
+        max_depth_new_edges = ave_width + 1.5
+    if max_width is None:
+        max_width = 1.5 + ave_width * math.log(ave_width + 1)
+    if max_height is None:
+        max_height = len(dsk)
+
+    # TODO: accept `dependencies=` keyword and return updated dependencies
+    deps = {k: get_dependencies(dsk, k, as_list=True) for k in dsk}
```

Comment:
Agreed.  This should behave the same as `fuse`.  This may be cheap to add now.  Give me a minute.